rasa/core/actions/forms.py

```diff
@@ -632,7 +632,24 @@ async def do(
         events_so_far: List[Event],
     ) -> List[Event]:
         """Executes form loop after activation."""
-        events = await self._validate_if_required(tracker, domain, output_channel, nlg)
+        events: List[Event] = []
+        """
+        Call to validation is not requried when the slots are already validated
+        at the time of form activation.
+        events_so_far:
+            - empty when slots have not been validated.
+            - has SlotSet objects when already validated.
+            - ActiveLoop object when events have not been validated.
+        Hence the events are filtered to remove ActiveLoop object that was added
+        at the time of form activation.
+        """
+        filtered_events = [
+            event for event in events_so_far if not isinstance(event, ActiveLoop)
+        ]
+        if not filtered_events:
```

Comment:
Not sure about this condition. Could it be events of other types (besides `ActiveLoop` and `SlotSet`)? If yes, maybe it would be better to check explicitly for presence of `SlotSet` in the list, rather than for absence of `ActiveLoop`?

Another note, that iterator could be used to avoid explicit list creation, something like:

```
if not any(isinstance(e, SlotSet) for e in events_so_far):
  # do validation
```

Reply:
At the time of form activation, the `ActiveLoop` object is added to the events and this is only available when `do` method is called immediately after the `activate` method. The goal is to not call the `validate` method under the `do` method at the time of form activation.

Reply:
The do method can take the following values,

- `events` from form activation, including the `ActiveLoop` object [`do` method should not validate]
- empty `events` when generally called. [`do` method must validate]
- `ActiveLoop` object from form activation, when a custom slot validation is present. [`do` method must validate]
Hence, we remove the `ActiveLoop` object and check for empty `events` to run the validation.

Reply:
Ok, thanks! It is clear now